core/src/main/scala/org/apache/spark/deploy/PythonRunner.scala

```diff
@@ -153,4 +154,30 @@ object PythonRunner {
       .map { p => formatPath(p, testWindows) }
   }
 
+  /**
+   * Resolves the ".py" files. ".py" file should not be added as is because PYTHONPATH does
+   * not expect a file. This method creates a temporary directory and puts the ".py" files
+   * if exist in the given paths.
+   */
+  private def resolvePyFiles(pyFiles: Array[String]): Array[String] = {
+    val dest = Utils.createTempDir(namePrefix = "localPyFiles")
+    pyFiles.flatMap { pyFile =>
+      // In case of client with submit, the python paths should be set before context
+      // initialization because the context initialization can be done later.
+      // We will copy the local ".py" files because ".py" file shouldn't be added
+      // alone but its parent directory in PYTHONPATH. See SPARK-24384.
+      if (pyFile.endsWith(".py")) {
+        val source = new File(pyFile)
+        if (source.exists() && source.canRead) {
```

Comment:
`source.isFile() && source.canRead()`

re: unreadable files, is there a check for it anywhere else? If not, that should be added, or the app might fail with some hard to debug exception.

Reply:
@vanzin, do you mean that this should be checked ahead (for example in SparkSubmit) before we are in this logic?

Just for clarification, this is just a sanity check. The previous behaviour was that the path is added but it's ignored and the current behaviour is that it doesn't add the path.

Reply:
I think providing a non-existent file to spark-submit should result in an error. Whether the error happens here or somewhere else it doesn't really matter.